core/templates/filters/translate.pipe.ts

```diff
@@ -0,0 +1,149 @@
+ */
+
+import { ChangeDetectorRef, OnDestroy, Pipe, PipeTransform }
+  from '@angular/core';
+import { Subscription } from 'rxjs';
+import { LangChangeEvent, TranslateService } from 'services/translate.service';
+import { UtilsService } from 'services/utils.service';
+
+/**
+ * Commonly used terms in this file.
+ * key - Key for i18n translation
+ * interpolateParams or params - Params for interpolation
+ * interpolatedValue - The final translation that is returned.
+ * Note: Intentionally left out the L of innerHTM"L" (only in this file) to
+ * avoid the linting errors.
+ * Example: <h1 [innerHTM]="'I18N_ABOUT_PAGE_HEADING' | translate:{x: 'val'}">
+ * 'I18N_ABOUT_PAGE_HEADING' is referred here as key.
+ * "translate" is the pipe. Every pipe must have a transform function. The
+ * transform function is called when angular encouters the pipe in HTML.
+ * The object following the pipe, i.e. {x: 'val'}, is another argument to the
+ * transform function. This object is called params or interpolationParams.
+ * Note: For every "| translate" angular encounters in the html, it creates a
+ * seperate instance of this pipe.
+ */
+
+@Pipe({
+  name: 'translate',
+  // This option is required to update the interpolatedValue when promises are
+  // resolved.
+  pure: false
+})
+export class TranslatePipe implements PipeTransform, OnDestroy {
+  interpolatedValue: string = '';
+  /**
+   * Variable to keep track of the last key sent for translation. If the
+   * previously sent key and params are same as the current one, the pipe
+   * doesn't request for translation and interpolation again.
+   */
+  lastKey: string;
+  /**
+   * Variable to keep track of the last param sent for translation. If the
+   * previously sent key and params are same as the current one, the pipe
+   * doesn't request for translation and interpolation again.
+   */
+  lastParams: Object;
+  onLangChange: Subscription;
+
+  constructor(
+    private translateService: TranslateService,
+    private ref: ChangeDetectorRef,
+    private utilsService: UtilsService
+  ) {}
+
+  /**
+   * Updates the value of interpolatedValue.
+   * @param {string} key - key for i18n translation
+   * @param {Object} interpolateParams - Params for interpolation
+   */
+  updateInterpolatedValue(key: string, interpolateParams?: Object): void {
+    this.interpolatedValue = this.translateService.getInterpolatedString(
+      key, interpolateParams);
+
+    // Storing the key to check if the key is same when the transform is invoked
+    // again.
+    this.lastKey = key;
+
+    // Storing the params to check if the params are same when the transform is
+    // invoked again.
+    this.lastParams = interpolateParams;
+    this.ref.markForCheck();
+  }
+
+  /**
+   * @param {string} key - key for i18n
+   * @param {Object} params - params for interpolation
+   * @returns {string} - Interpolated I18n value
+   */
+  transform(key: string, params?: Object | undefined): string {
```

Comment:
Can you please pick a more descriptive name? ("transform" is about as vague as one can get...)

Reply:
Ah, I see. Could you please make claer in the docstring (and other docstrings, where applicable) that this function overrides/implements an existing interface function? One of the issues with trying to understand what is going on is that there is no actual docstring explaining what this function does -- see [this comment thread](https://github.com/oppia/oppia/pull/9398/files#r435994565) above. **Please do this for all JSDocs your PR is touching.**

Reply:
Done!